### PR TITLE
Removed unused _RequiredParameter

### PR DESCRIPTION
**Pull request by dmpiergiacomo**
As per this [discussion](https://discuss.pytorch.org/t/a-question-about-requiredparameter/137977), I figured that `_RequiredParameter` is no longer used.

The `required` object was initially introduced in this [PR](https://github.com/pytorch/pytorch/commit/4db66679238dae8539c270a61f60b9c0c4bb440d) as the `SGD` optimizer did not offer a default value for the learning rate. However there isn't a single place in the code base using `_RequiredParameter`, nor `required`. I am therefore removing unused `_RequiredParameter` and `required`.

Everything not included in this PR is Not a Contribution.

### CODE REVIEW

**Review by janeyx99 — DISMISSED** [review]
thanks! we did indeed stop using this after SGD got a default

**Comment by dmpiergiacomo** [comment]
@janeyx99 and @cyyever, could you please share some insights on why this PR has been closed and not merged? It seems like the `try-merge` job went through.

Thanks!

**Comment by cyyever** [comment]
It's merged.

**Comment by malfet** [comment]
> @janeyx99 and @cyyever, could you please share some insights on why this PR has been closed and not merged? It seems like the `try-merge` job went through.

@dmpiergiacomo it's been merged, see https://github.com/pytorch/pytorch/commit/6a5f895e549665a6895c84881a35736677071048, but it caused a lots of regressions in inductor test, see https://hud.pytorch.org/hud/pytorch/pytorch/7c52c97a65f58e1de2967509ab732e20f468dae8/1?per_page=50&name_filter=cpu_inductor_torchbench

**Comment by dmpiergiacomo** [comment]
Thanks @malfet! I'm not familiar with Pytorch CI and integration tests. Is there any action I should take, or should I simply wait for all the new checks to complete?

**Comment by janeyx99** [comment]
@dmpiergiacomo It looks like this a BC breaking change for the timm models tested https://hud.pytorch.org/pytorch/pytorch/commit/6a5f895e549665a6895c84881a35736677071048

The action item would be to figure out where these models are defined, fix them, and then have a way of being confident that no one else is using `required`.  This may not be as trivial of a change as expected. 😬

**Comment by janeyx99** [comment]
Oh in fact, this is not okay to be removed as simply as here because `required` is a public API we've exposed. :/ It would have to go through a deprecation cycle at the very least if we want to deprecate it, which is unlikely given that others rely on this API and it does not cost too much to maintain.

**Comment by dmpiergiacomo** [comment]
Thanks for the explanation @janeyx99! I understand the complication. Out of curiosity, how do you track which APIs are used and which aren't? Is this something you were able to automate? I assume Pytorch doesn't have any backdoor, but maybe you figure out a way?

**Comment by janeyx99** [comment]
haha it isn't automated--I normally do a GitHub search for public usage, but searching for `required` will return false positives. Though it'd be fine to  search for some regex for `from torch\.optim.*import.*required` or `Optimizer.required` or something.

In this case, I didn't do a check, as we cannot remove any public APIs anyway without doing a deprecation cycle. I'm willing to be convinced that `required` is worth deprecating, but there'd need to be a good reason to break people's current code (remember that GH is only public usage, so we don't know who privately uses it).

**Comment by dmpiergiacomo** [comment]
That's very clear. Thanks! I guess I could scrape a bit of GitHub in regex style. I'll probably do it in as soon as I have more capacity next week.

The strict deprecation cycle you talk about justifies well why I found multiple pieces that aren't used in the repo. For example some small things also in the backward engine algorithms.
